tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,290 @@
+        atomic_number : int
+            Current atomic number
+        ion_number : int
+            Current ion number
+        radiative_recombination_rate : DataFrame
+            Rad. recomb. rate for current atomic number
+        photo_ion_rates : DataFrame
+            Photo ion. rate for current atomic number
+        coll_ion_rate : _type_
+            Coll. ion. rate for current atomic number
+        coll_recomb_rate : _type_
+            Coll. recomb. rate for current atomic number
+
+        Returns
+        -------
+        numpy.array
+            Rate matrix block with a changed row for NLTE ionization treatment
+        """
+        ion_rates = photo_ion_rates + coll_ion_rate
+        recomb_rate = radiative_recombination_rate + coll_recomb_rate
+        if atomic_number == ion_number:
+            rate_matrix_block[ion_number, :] = 1.0
+        else:
+            ion_rate_matrix = NLTERateEquationSolver.ion_matrix(
+                ion_rates, atomic_number
+            )
+            recomb_rate_matrix = NLTERateEquationSolver.recomb_matrix(
+                recomb_rate, atomic_number
+            )
+            rate_matrix_block[ion_number, :] = (
+                ion_rate_matrix + recomb_rate_matrix
+            )[ion_number, :]
+        return rate_matrix_block
+
+    @staticmethod
+    def lte_rate_matrix_block(phi_block, electron_density):
+        """Creates the generic LTE block for rate matrix.
+
+        Parameters
+        ----------
+        phi_block : DataFrame
+            Saha Factors for current atomic number
+        electron_density : float
+            Current guess for electron density
+
+        Returns
+        -------
+        numpy.array
+            LTE block for rate matrix
+        """
+        lte_rate_vector_block = -1.0 * np.hstack([*phi_block.values, -1.0])
+        lte_rate_matrix_block = np.diag(lte_rate_vector_block)
+        n_e_initial = np.ones(len(phi_block)) * electron_density
+        n_e_matrix = np.diag(n_e_initial, 1)
+        lte_rate_matrix_block += n_e_matrix
+        lte_rate_matrix_block[-1, :] = 1.0
+        return lte_rate_matrix_block
+
+    @staticmethod
+    def prepare_phi(phi):
+        phi[phi == 0.0] = 1.0e-10 * phi[phi > 0.0].min().min()
+        return phi
+
+    @staticmethod
+    def recomb_matrix(recomb_rate, atomic_number):
+        offdiag = np.zeros(atomic_number)
+        index = recomb_rate.index
+        for i in index:
+            offdiag[i] = recomb_rate.loc[i]
+        diag = np.hstack([np.zeros(1), -offdiag])
+        return np.diag(diag) + np.diag(offdiag, k=1)
+
+    @staticmethod
+    def ion_matrix(ion_rate, atomic_number):
+        offdiag = np.zeros(atomic_number)
+        index = ion_rate.index
+        for i in index:
+            offdiag[i] = ion_rate.loc[i]
+        diag = np.hstack([-offdiag, np.zeros(1)])
+        return np.diag(diag) + np.diag(offdiag, k=-1)
+
+    @staticmethod
+    def prepare_last_row(atomic_numbers):
```

Comment:
I would consider renaming `last_row` to `charge_conservation_row`. This would make the meaning of the row clearer.

Reply:
Done